jsprit-core/src/main/java/com/graphhopper/jsprit/core/problem/job/AbstractJob.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to GraphHopper GmbH under one or more contributor
+ * license agreements. See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ *
+ * GraphHopper GmbH licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.graphhopper.jsprit.core.problem.job;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import com.graphhopper.jsprit.core.problem.Location;
+import com.graphhopper.jsprit.core.problem.solution.route.activity.TimeWindow;
+
+/**
+ * Created by schroeder on 14.07.14.
+ */
+public abstract class AbstractJob implements Job {
+
+    private int index;
+
+    protected List<Location> allLocations = new ArrayList<>();
+
+    private JobActivityList activityList;
+
+    protected Set<TimeWindow> operationTimeWindows = new HashSet<>();
+
+    public AbstractJob() {
+        super();
+        activityList = new SequentialJobActivityList(this);
+    }
+
+    @Override
+    public int getIndex() {
+        return index;
+    }
+
+    public void setIndex(int index) {
+        this.index = index;
+    }
+
+    protected void addLocation(Location location) {
```

Comment:
As you can see, this is a protected method. This is used by the constructor (or to be correct, by the post process of the builder) to create the caches, which are used by the problem creation. 
Generally, the user defines Job implementations and then instantiate them (or the ones already in jsprit, sush as Shipment, Service, etc.).
These functions, and the caches they create makes the other part of the code possible to handle any type of Jobs, uniformly. For example, when the problem builder creates the activities, it doesn't care what type of job is it, it simply asks the list of activities from the job and adds them. With the locations and other information the case is the same. 
On the other hand, when activities are created all this information is copied to them, so they will stand as a complete data source and don't need to delegate to their jobs. It makes them also flexible: the job knows how to fill in the data from its own fields (done when jobActivityList is initialized), and later on, an activity doesn't need to know where these data came from. It makes activity handling uniform as well.